File: src/emulate.rs
use ::serde::Deserialize;
use magnus::{
    Error, Module, Object, RModule, Ruby, Value, function, method,
    typed_data::{Inspect, Obj},
};

use crate::options::{NativeOption, Options};

/// Keyword arguments accepted by `Wreq::Emulation.new`.
#[derive(Default, Deserialize)]
struct Builder {
    /// The browser profile to emulate.
    #[serde(default)]
    profile: NativeOption<Obj<Profile>>,

    /// The operating-system profile to emulate.
    #[serde(default)]
    platform: NativeOption<Obj<Platform>>,

    /// Whether HTTP/2 settings are emulated.
    http2: Option<bool>,

    /// Whether browser headers are emulated.
    headers: Option<bool>,
}

// ===== impl Builder =====

impl Builder {
    /// Deserialize and convert one validated emulation options Hash.
    ///
    /// # Errors
    ///
    /// Returns `ArgumentError` for unknown or duplicate options and `TypeError`
    /// for invalid option values.
    fn from_options(options: Options<'_>) -> Result<Self, Error> {
        let mut builder = options.validate_keys::<Self>()?.deserialize::<Self>()?;
        extract_native_option!(options, builder, profile);
        extract_native_option!(options, builder, platform);
        Ok(builder)
    }
}

define_ruby_enum!(
    /// An emulation profile.
    const,
    Profile,
    "Wreq::Profile",
    wreq_util::Profile,
    Chrome100,
    Chrome101,
    Chrome104,
    Chrome105,
    Chrome106,
    Chrome107,
    Chrome108,
    Chrome109,
    Chrome110,
    Chrome114,
    Chrome116,
    Chrome117,
    Chrome118,
    Chrome119,
    Chrome120,
    Chrome123,
    Chrome124,
    Chrome126,
    Chrome127,
    Chrome128,
    Chrome129,
    Chrome130,
    Chrome131,
    Chrome132,
    Chrome133,
    Chrome134,
    Chrome135,
    Chrome136,
    Chrome137,
    Chrome138,
    Chrome139,
    Chrome140,
    Chrome141,
    Chrome142,
    Chrome143,
    Chrome144,
    Chrome145,
    Chrome146,
    Chrome147,
    Chrome148,
    Chrome149,

    Edge101,
    Edge122,
    Edge127,
    Edge131,
    Edge134,
    Edge135,
    Edge136,
    Edge137,
    Edge138,
    Edge139,
    Edge140,
    Edge141,
    Edge142,
    Edge143,
    Edge144,
    Edge145,
    Edge146,
    Edge147,
    Edge148,

    Firefox109,
    Firefox117,
    Firefox128,
    Firefox133,
    Firefox135,
    FirefoxPrivate135,
    FirefoxAndroid135,
    Firefox136,
    FirefoxPrivate136,
    Firefox139,
    Firefox142,
    Firefox143,
    Firefox144,
    Firefox145,
    Firefox146,
    Firefox147,
    Firefox148,
    Firefox149,
    Firefox150,
    Firefox151,

    SafariIos17_2,
    SafariIos17_4_1,
    SafariIos16_5,
    Safari15_3,
    Safari15_5,
    Safari15_6_1,
    Safari16,
    Safari16_5,
    Safari17_0,
    Safari17_2_1,
    Safari17_4_1,
    Safari17_5,
    Safari17_6,
    Safari18,
    SafariIPad18,
    Safari18_2,
    Safari18_3,
    Safari18_3_1,
    SafariIos18_1_1,
    Safari18_5,
    Safari26,
    Safari26_1,
    Safari26_2,
    Safari26_3,
    Safari26_4,
    SafariIos26,
    SafariIos26_2,
    SafariIPad26,
    SafariIpad26_2,

    OkHttp3_9,
    OkHttp3_11,
    OkHttp3_13,
    OkHttp3_14,
    OkHttp4_9,
    OkHttp4_10,
    OkHttp4_12,
    OkHttp5,

    Opera116,
    Opera117,
    Opera118,
    Opera119,
    Opera120,
    Opera121,
    Opera122,
    Opera123,
    Opera124,
    Opera125,
    Opera126,
    Opera127,
    Opera128,
    Opera129,
    Opera130,
    Opera131,
);

define_ruby_enum!(
    /// An emulation profile for OS.
    const,
    Platform,
    "Wreq::Platform",
    wreq_util::Platform,
    Windows,
    MacOS,
    Linux,
    Android,
    IOS,
);

/// A struct to represent the `EmulationOption` class.
#[derive(Clone)]
#[magnus::wrap(class = "Wreq::Emulation", free_immediately, size)]
pub struct Emulation(pub wreq_util::Emulation);

// ===== impl Profile =====

impl Profile {
    pub fn to_s(&self) -> String {
        self.into_ffi().inspect()
    }
}

// ===== impl Platform =====

impl Platform {
    pub fn to_s(&self) -> String {
        self.into_ffi().inspect()
    }
}

// ===== impl Emulation =====

impl Emulation {
    /// Create emulation settings from one optional Hash.
    ///
    /// Unknown keys, non-Hash arguments, and extra positional arguments are
    /// rejected before the native emulation value is built.
    ///
    /// # Errors
    ///
    /// Returns `ArgumentError` for unknown keys or argument count and
    /// `TypeError` for invalid option values.
    fn new(ruby: &Ruby, args: &[Value]) -> Result<Self, Error> {
        let mut params = Options::from_args(ruby, args, "emulation")?
            .map(Builder::from_options)
            .transpose()?
            .unwrap_or_default();

        let emulation = wreq_util::Emulation::builder()
            .profile(
                params
                    .profile
                    .take()
                    .map(|obj| obj.into_ffi())
                    .unwrap_or_default(),
            )
            .platform(
                params
                    .platform
                    .take()
                    .map(|os| os.into_ffi())
                    .unwrap_or_default(),
            )
            .http2(params.http2.unwrap_or(true))
            .headers(params.headers.unwrap_or(true))
            .build();

        Ok(Self(emulation))
    }
}

pub fn include(ruby: &Ruby, gem_module: &RModule) -> Result<(), Error> {
    // Profile enum binding
    let profile = gem_module.define_class("Profile", ruby.class_object())?;
    profile.define_method("to_s", method!(Profile::to_s, 0))?;
    Profile::define_constants(profile)?;

    // Platform enum binding
    let platform = gem_module.define_class("Platform", ruby.class_object())?;
    platform.define_method("to_s", method!(Platform::to_s, 0))?;
    Platform::define_constants(platform)?;

    // Emulation class binding
    let emulation = gem_module.define_class("Emulation", ruby.class_object())?;
    emulation.define_singleton_method("new", function!(Emulation::new, -1))?;
    Ok(())
}

File: lib/wreq_ruby/emulate.rb
# frozen_string_literal: true

# Profile and platform constants mirror the native enum variant names.
# standard:disable Naming/ConstantName

module Wreq
  # Browser and client fingerprint profile enumeration backed by Rust.
  #
  # Variants are exposed as constants under this class.
  # Each constant is an instance of {Wreq::Profile} and can be passed to
  # {Wreq::Emulation.new} via the `profile:` keyword.
  #
  # @example Using a predefined profile
  #   profile = Wreq::Profile::Chrome117
  #   profile.class #=> Wreq::Profile
  #
  # @example Applying a profile to emulation
  #   emu = Wreq::Emulation.new(profile: Wreq::Profile::Chrome117)
  class Profile
    # Constants are set by the native extension at initialization.
    # These stubs are for documentation only.
    unless const_defined?(:Chrome100)
      Chrome100 = nil
      Chrome101 = nil
      Chrome104 = nil
      Chrome105 = nil
      Chrome106 = nil
      Chrome107 = nil
      Chrome108 = nil
      Chrome109 = nil
      Chrome110 = nil
      Chrome114 = nil
      Chrome116 = nil
      Chrome117 = nil
      Chrome118 = nil
      Chrome119 = nil
      Chrome120 = nil
      Chrome123 = nil
      Chrome124 = nil
      Chrome126 = nil
      Chrome127 = nil
      Chrome128 = nil
      Chrome129 = nil
      Chrome130 = nil
      Chrome131 = nil
      Chrome132 = nil
      Chrome133 = nil
      Chrome134 = nil
      Chrome135 = nil
      Chrome136 = nil
      Chrome137 = nil
      Chrome138 = nil
      Chrome139 = nil
      Chrome140 = nil
      Chrome141 = nil
      Chrome142 = nil
      Chrome143 = nil
      Chrome144 = nil
      Chrome145 = nil
      Chrome146 = nil
      Chrome147 = nil
      Chrome148 = nil
      Chrome149 = nil
      Chrome150 = nil

      Edge101 = nil
      Edge122 = nil
      Edge127 = nil
      Edge131 = nil
      Edge134 = nil
      Edge135 = nil
      Edge136 = nil
      Edge137 = nil
      Edge138 = nil
      Edge139 = nil
      Edge140 = nil
      Edge141 = nil
      Edge142 = nil
      Edge143 = nil
      Edge144 = nil
      Edge145 = nil
      Edge146 = nil
      Edge147 = nil
      Edge148 = nil

      Firefox109 = nil
      Firefox117 = nil
      Firefox128 = nil
      Firefox133 = nil
      Firefox135 = nil
      FirefoxPrivate135 = nil
      FirefoxAndroid135 = nil
      Firefox136 = nil
      FirefoxPrivate136 = nil
      Firefox139 = nil
      Firefox142 = nil
      Firefox143 = nil
      Firefox144 = nil
      Firefox145 = nil
      Firefox146 = nil
      Firefox147 = nil
      Firefox148 = nil
      Firefox149 = nil
      Firefox150 = nil
      Firefox151 = nil

      SafariIos17_2 = nil
      SafariIos17_4_1 = nil
      SafariIos16_5 = nil
      Safari15_3 = nil
      Safari15_5 = nil
      Safari15_6_1 = nil
      Safari16 = nil
      Safari16_5 = nil
      Safari17_0 = nil
      Safari17_2_1 = nil
      Safari17_4_1 = nil
      Safari17_5 = nil
      Safari17_6 = nil
      Safari18 = nil
      SafariIPad18 = nil
      Safari18_2 = nil
      Safari18_3 = nil
      Safari18_3_1 = nil
      SafariIos18_1_1 = nil
      Safari18_5 = nil
      Safari26 = nil
      Safari26_1 = nil
      Safari26_2 = nil
      Safari26_3 = nil
      Safari26_4 = nil
      SafariIos26 = nil
      SafariIos26_2 = nil
      SafariIPad26 = nil
      SafariIpad26_2 = nil

      OkHttp3_9 = nil
      OkHttp3_11 = nil
      OkHttp3_13 = nil
      OkHttp3_14 = nil
      OkHttp4_9 = nil
      OkHttp4_10 = nil
      OkHttp4_12 = nil
      OkHttp5 = nil

      Opera116 = nil
      Opera117 = nil
      Opera118 = nil
      Opera119 = nil
      Opera120 = nil
      Opera121 = nil
      Opera122 = nil
      Opera123 = nil
      Opera124 = nil
      Opera125 = nil
      Opera126 = nil
      Opera127 = nil
      Opera128 = nil
      Opera129 = nil
      Opera130 = nil
      Opera131 = nil
    end

    unless method_defined?(:to_s)
      # Returns the profile name.
      # @return [String] Profile name as a string
      def to_s
      end
    end

    unless method_defined?(:==)
      # Value-based equality.
      # @param other [Object]
      # @return [Boolean]
      def ==(other)
      end
    end

    unless method_defined?(:eql?)
      # Strict equality for Hash key and Set member semantics.
      # @param other [Object]
      # @return [Boolean]
      def eql?(other)
      end
    end

    unless method_defined?(:hash)
      # Hash value consistent with {#eql?} for use as Hash keys.
      # @return [Integer]
      def hash
      end
    end
  end

  # Operating system platform enumeration backed by Rust.
  #
  # Variants are exposed as constants under this class.
  # Each constant is an instance of {Wreq::Platform} and can be passed to
  # {Wreq::Emulation.new} via the `platform:` keyword.
  #
  # @example Using a predefined platform
  #   platform = Wreq::Platform::Windows
  #   platform.class #=> Wreq::Platform
  #
  # @example Applying a platform to emulation
  #   emu = Wreq::Emulation.new(platform: Wreq::Platform::Windows)
  class Platform
    # Constants are set by the native extension at initialization.
    # These stubs are for documentation only.
    unless const_defined?(:Windows)
      Windows = nil
      MacOS = nil
      Linux = nil
      Android = nil
      IOS = nil
    end

    unless method_defined?(:to_s)
      # Returns the platform name.
      # @return [String] Platform name as a string
      def to_s
      end
    end

    unless method_defined?(:to_sym)
      # Returns the platform as a lowercase symbol (e.g. :windows, :linux).
      # @return [Symbol]
      def to_sym
      end
    end

    unless method_defined?(:==)
      # Value-based equality.
      # @param other [Object]
      # @return [Boolean]
      def ==(other)
      end
    end

    unless method_defined?(:eql?)
      # Strict equality for Hash key and Set member semantics.
      # @param other [Object]
      # @return [Boolean]
      def eql?(other)
      end
    end

    unless method_defined?(:hash)
      # Hash value consistent with {#eql?} for use as Hash keys.
      # @return [Integer]
      def hash
      end
    end
  end

  # Emulation option wrapper.
  #
  # This class combines a fingerprint `profile`, an OS `platform`, and toggles
  # for HTTP/2 and automatic default headers. The actual implementation is
  # provided by Rust.
  #
  # `profile:` defaults to the library's default profile when omitted.
  # `platform:` defaults to the library's default platform when omitted.
  #
  # @example Create an emulation option
  #   emu = Wreq::Emulation.new(
  #     profile: Wreq::Profile::Chrome117,
  #     platform: Wreq::Platform::Windows,
  #     http2: true,
  #     headers: true
  #   )
  #
  # @param profile [Wreq::Profile, nil] Fingerprint profile to emulate
  # @param platform [Wreq::Platform, nil] Operating system platform to emulate
  # @param http2 [Boolean, nil] Whether HTTP/2 emulation is enabled; defaults
  #   to true when omitted or nil
  # @param headers [Boolean, nil] Whether default emulation headers are enabled;
  #   defaults to true when omitted or nil
  # @return [Wreq::Emulation] Configured emulation settings
  # @raise [ArgumentError] if an option is unknown or extra arguments are given
  # @raise [TypeError] if the option argument is not a Hash or a value has the
  #   wrong Ruby type
  class Emulation
    # Native fields and methods are set by the extension.
    # This stub is for documentation only.
    unless singleton_methods(false).include?(:new)
      # @param profile [Wreq::Profile, nil] Fingerprint profile to emulate
      # @param platform [Wreq::Platform, nil] Operating system platform to emulate
      # @param http2 [Boolean, nil] Whether HTTP/2 emulation is enabled; defaults
      #   to true when omitted or nil
      # @param headers [Boolean, nil] Whether default emulation headers are enabled;
      #   defaults to true when omitted or nil
      # @return [Wreq::Emulation] Configured emulation settings
      # @raise [ArgumentError] if an option is unknown or extra arguments are given
      # @raise [TypeError] if an option or value has the wrong Ruby type
      def self.new(**options)
      end
    end
  end
end

# standard:enable Naming/ConstantName
